File: class27/python/PES2701-Documentacao.py
"""
PES2701 - Geração Automática de Documentação (Docstrings)
Aula 27: Documentação Técnica e Projeto Integrador Final
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Demonstra documentação inline com docstrings formatadas e extração
automática de informações para documentação, usando como exemplo o
cálculo de locação de veículos do Projeto Integrador (Sistema Locadora).
"""

import inspect


class Veiculo:
    """Classe base para veículos disponíveis na locadora.

    Attributes:
        placa: Placa de identificação do veículo.
        modelo: Nome/modelo do veículo.
        valor_diaria: Valor cobrado por dia de locação.
    """

    def __init__(self, placa, modelo, valor_diaria):
        self.placa = placa
        self.modelo = modelo
        self.valor_diaria = valor_diaria

    def calcular_locacao(self, dias):
        """Calcula o valor da locação deste veículo para um número de dias.

        Args:
            dias: Número de dias da locação.

        Returns:
            float: Valor total (valor_diaria * dias).
        """
        return self.valor_diaria * dias


class Carro(Veiculo):
    """Veículo do tipo carro, com número de portas."""

    def __init__(self, placa, modelo, valor_diaria, portas):
        super().__init__(placa, modelo, valor_diaria)
        self.portas = portas


class Moto(Veiculo):
    """Veículo do tipo moto, com cilindrada."""

    def __init__(self, placa, modelo, valor_diaria, cilindrada):
        super().__init__(placa, modelo, valor_diaria)
        self.cilindrada = cilindrada


def calcular_locacao(veiculo: Veiculo, dias: int) -> float:
    """
    Calcula o valor total da locação de um veículo.

    Args:
        veiculo: Instância de Veiculo (Carro ou Moto).
        dias: Número de dias da locação (deve ser >= 1).

    Returns:
        float: Valor total da locação em reais.

    Raises:
        ValueError: Se dias < 1.
        TypeError: Se veiculo não for instância de Veiculo.

    Example:
        >>> carro = Carro("ABC-1234", "Gol", 100.0, 4)
        >>> calcular_locacao(carro, 5)
        500.0
    """
    if not isinstance(veiculo, Veiculo):
        raise TypeError("veiculo deve ser uma instância de Veiculo")
    if dias < 1:
        raise ValueError("dias deve ser >= 1")
    return veiculo.calcular_locacao(dias)


def extrair_docstrings(modulo_ou_classe):
    """
    Extrai docstrings de todas as funções e métodos públicos
    de um módulo ou classe.

    Args:
        modulo_ou_classe: Módulo ou classe a ser inspecionado.

    Returns:
        List[dict]: Lista com nome, docstring e assinatura de cada callable.
    """
    resultado = []
    for nome, obj in inspect.getmembers(modulo_ou_classe):
        if nome.startswith("_"):
            continue
        if callable(obj):
            try:
                sig = str(inspect.signature(obj))
            except (ValueError, TypeError):
                sig = "()"
            doc = inspect.getdoc(obj) or "(sem documentação)"
            resultado.append({"nome": nome, "assinatura": sig, "docstring": doc})
    return resultado


if __name__ == "__main__":
    import __main__

    print("=" * 60)
    print("  DOCUMENTAÇÃO EXTRAÍDA AUTOMATICAMENTE")
    print("=" * 60)

    carro = Carro("ABC-1234", "Gol", 100.0, 4)
    moto = Moto("XYZ-5678", "CG 160", 60.0, 160)

    print(f"\nLocação do carro por 5 dias: R$ {calcular_locacao(carro, 5):.2f}")
    print(f"Locação da moto por 3 dias: R$ {calcular_locacao(moto, 3):.2f}")

    docs = extrair_docstrings(__main__)
    for item in docs:
        print(f"\n--- {item['nome']}{item['assinatura']} ---")
        print(f"  {item['docstring']}")

    print(f"\n✓ {len(docs)} funções documentadas.")
    print("✓ Docstrings bem escritas são a base de uma boa documentação técnica.")

File: class27/java/PES2701-Documentacao.java
/*
 * PES2701 - Geração de Documentação (Javadoc)
 * Aula 27: Documentação Técnica e Projeto Integrador Final
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Demonstra documentação inline com Javadoc, usando como exemplo o
 * cálculo de locação de veículos do Projeto Integrador (Sistema Locadora).
 *
 * Compilar/executar:
 *   javac PES2701-Documentacao.java && java Documentacao
 */

/** Classe base para veículos disponíveis na locadora. */
class Veiculo {
    String placa;
    String modelo;
    double valorDiaria;

    Veiculo(String placa, String modelo, double valorDiaria) {
        this.placa = placa;
        this.modelo = modelo;
        this.valorDiaria = valorDiaria;
    }

    /**
     * Calcula o valor da locação deste veículo para um número de dias.
     *
     * @param dias número de dias da locação
     * @return valor total (valorDiaria * dias)
     */
    double calcularLocacao(int dias) {
        return valorDiaria * dias;
    }
}

/** Veículo do tipo carro, com número de portas. */
class Carro extends Veiculo {
    int portas;

    Carro(String placa, String modelo, double valorDiaria, int portas) {
        super(placa, modelo, valorDiaria);
        this.portas = portas;
    }
}

/** Veículo do tipo moto, com cilindrada. */
class Moto extends Veiculo {
    int cilindrada;

    Moto(String placa, String modelo, double valorDiaria, int cilindrada) {
        super(placa, modelo, valorDiaria);
        this.cilindrada = cilindrada;
    }
}

/**
 * Classe que demonstra boas práticas de documentação com Javadoc.
 *
 * <p>Uma boa documentação de código inclui:
 * <ul>
 *   <li>Descrição do propósito da classe/método</li>
 *   <li>Parâmetros com @param</li>
 *   <li>Retorno com @return</li>
 *   <li>Exceções com @throws</li>
 * </ul>
 *
 * @author Prof. Giovane Barcelos
 * @version 1.0
 */
class Documentacao {

    /**
     * Calcula o valor total da locação de um veículo.
     *
     * @param veiculo  instância de Veiculo (Carro ou Moto)
     * @param dias     número de dias da locação (deve ser >= 1)
     * @return         valor total da locação em reais
     * @throws IllegalArgumentException se dias < 1
     * @throws NullPointerException     se veiculo for null
     *
     * <pre>{@code
     *   Carro carro = new Carro("ABC-1234", "Gol", 100.0, 4);
     *   double total = calcularLocacao(carro, 5); // 500.0
     * }</pre>
     */
    public static double calcularLocacao(Veiculo veiculo, int dias) {
        if (veiculo == null) {
            throw new NullPointerException("veiculo não pode ser null");
        }
        if (dias < 1) {
            throw new IllegalArgumentException("dias deve ser >= 1");
        }
        return veiculo.calcularLocacao(dias);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  DOCUMENTAÇÃO COM JAVADOC");
        System.out.println("=".repeat(60));

        Carro carro = new Carro("ABC-1234", "Gol", 100.0, 4);
        Moto moto = new Moto("XYZ-5678", "CG 160", 60.0, 160);

        System.out.printf("%n  Locação do carro por 5 dias: R$ %.2f%n", calcularLocacao(carro, 5));
        System.out.printf("  Locação da moto por 3 dias: R$ %.2f%n", calcularLocacao(moto, 3));

        System.out.println("\n✓ Javadoc: documentação integrada ao código.");
        System.out.println("✓ Gere HTML com: javadoc -d docs *.java");
    }
}
